src/lib/drivers/smbus_sbs/SBS.hpp

```diff
@@ -0,0 +1,325 @@
+#define BATT_SMBUS_AVERAGE_TIME_TO_EMPTY                0x12            ///< predicted remaining battery capacity based on the present rate of discharge in min
+#define BATT_SMBUS_CYCLE_COUNT                          0x17            ///< number of cycles the battery has experienced
+#define BATT_SMBUS_DESIGN_CAPACITY                      0x18            ///< design capacity register
+#define BATT_SMBUS_DESIGN_VOLTAGE                       0x19            ///< design voltage register
+#define BATT_SMBUS_MANUFACTURER_NAME                    0x20            ///< manufacturer name
+#define BATT_SMBUS_MANUFACTURER_NAME_SIZE               21              ///< manufacturer name data size (not part of SBS spec)
+#define BATT_SMBUS_MANUFACTURE_DATE                     0x1B            ///< manufacture date register
+#define BATT_SMBUS_SERIAL_NUMBER                        0x1C            ///< serial number register
+#define BATT_SMBUS_MANUFACTURER_ACCESS                  0x00
+#define BATT_SMBUS_MANUFACTURER_DATA                    0x23
+
+template<class T>
+class SMBUS_SBS_BaseClass : public I2CSPIDriver<T>
+{
+public:
+	SMBUS_SBS_BaseClass(I2CSPIBusOption bus_option, const int bus, SMBus *interface);
+	SMBUS_SBS_BaseClass();
+
+	~SMBUS_SBS_BaseClass();
+
+	static void print_usage();
+
+	friend SMBus;
+
+	int populate_smbus_data(battery_status_s &msg);
+
+	virtual void RunImpl(); // Can be overriden by derived implimentation
+
+	virtual void custom_method(const BusCLIArguments &cli) = 0; //Has be overriden by derived implimentation
+
+	/**
+	 * @brief Returns the SBS serial number of the battery device.
+	 * @return Returns the SBS serial number of the battery device.
+	 */
+	uint16_t get_serial_number();
+
+	/**
+	* @brief Read info from battery on startup.
+	* @return Returns PX4_OK on success, PX4_ERROR on failure. Can be overriden by derived implimentation
+	*/
+	virtual int get_startup_info();
+
+	/**
+	 * @brief Gets the SBS manufacture date of the battery.
+	 * @return Returns PX4_OK on success, PX4_ERROR on failure.
+	 */
+	int manufacture_date();
+
+	/**
+	 * @brief Gets the SBS manufacturer name of the battery device.
+	 * @param manufacturer_name Pointer to a buffer into which the manufacturer name is to be written.
+	 * @param max_length The maximum number of bytes to attempt to read from the manufacturer name register,
+	 *                   including the null character that is appended to the end.
+	 * @return Returns PX4_OK on success, PX4_ERROR on failure.
+	 */
+	int manufacturer_name(uint8_t *manufacturer_name, const uint8_t length);
+
+	/**
+	 * @brief Enables or disables the cell under voltage protection emergency shut off.
+	 */
+
+	void suspend();
+
+	void resume();
+
+protected:
+
+	SMBus *_interface;
+
+	perf_counter_t _cycle{perf_alloc(PC_ELAPSED, "batmon_cycle")}; // TODO
+
+	/** @param _batt_topic uORB battery topic. */
+	orb_advert_t _batt_topic{nullptr};
```

Comment:
This should probably be a `uORB::Publication<...>`

Reply:
why so ?

Reply:
Do you have a reference or document for this?